src/model/grid.rs

```diff
@@ -0,0 +1,217 @@
+use macroquad::prelude::{BLACK, draw_rectangle, screen_height, screen_width};
+
+pub struct Grid {
+    pub width: usize,
+    pub height: usize,
+    matrix: Vec<Vec<bool>>,
+}
+
+impl Grid {
+    pub fn new(width: usize, height: usize) -> Grid {
+        if width == 0 || height == 0 {
+            panic!("Dimensions for the grid can not be 0");
+        }
+
+        let matrix = vec![vec![false; width]; height];
+
+        Grid {
+            width,
+            height,
+            matrix,
+        }
+    }
+
+    pub fn draw(&self) {
+        let cell_width = screen_width() / self.width as f32;
+        let cell_height = screen_height() / self.height as f32;
+
+        for row_n in 0..self.width {
+            for col_n in 0..self.height {
+                let is_live = self.matrix[row_n][col_n];
+
+                if !is_live {
+                    // If not live don't draw.
+                    continue;
+                }
+
+                // If live, draw cell.
+                let x = row_n as f32 * cell_width;
+                let y = col_n as f32 * cell_height;
+
+                draw_rectangle(x, y, cell_width, cell_height, BLACK);
+            }
+        }
+    }
+
+    pub fn get(&self, x: usize, y: usize) -> bool {
+        self.matrix[x][y]
+    }
+
+    fn get_mut(&mut self, x: usize, y: usize) -> &mut bool {
+        &mut self.matrix[x][y]
+    }
+
+    pub fn update(&mut self, x: usize, y: usize, value: bool) {
+        *self.get_mut(x, y) = value;
+    }
+
+    pub fn toggle(&mut self, x: usize, y: usize) {
+        self.update(x, y, !self.get(x, y));
+    }
+
+    pub fn coordinates(&self) -> Vec<(usize, usize)> {
+        let mut coordinates = Vec::new();
+
+        for x in 0..self.width {
+            for y in 0..self.height {
+                coordinates.push((x, y));
+            }
+        }
+
+        coordinates
+    }
+
+    pub fn neighbours(&self, x: usize, y: usize) -> Vec<(usize, usize)> {
+        let mut neighbours = Vec::new();
+
+        for dx in -1i32..=1 {
+            for dy in -1i32..=1 {
+                let nx = x as i32 + dx;
+                let ny = y as i32 + dy;
+
+                let is_target_cell = (nx, ny) == (x as i32, y as i32);
+                let is_out_of_bounds =
+                    nx < 0 || nx >= self.width as i32 || ny < 0 || ny >= self.height as i32;
+
+                if is_target_cell || is_out_of_bounds {
+                    continue;
+                }
+
+                // Cast should not fail after checks
+                neighbours.push((nx as usize, ny as usize));
+            }
+        }
+
+        neighbours
+    }
+}
+
+impl Clone for Grid {
+    fn clone(&self) -> Self {
+        Grid {
+            width: self.width,
+            height: self.height,
+            matrix: self.matrix.clone(),
+        }
+    }
+}
```

Comment:
You can replace this `impl` block with a `derive` attribute in the `Grid` struct definition:

```rust
#[derive(Clone)]
pub struct Grid {
    pub width: usize,
    pub height: usize,
    matrix: Vec<Vec<bool>>,
}
```

Reply:
Thanks, I missed that. It is updated now.